File: packages/shared/src/constants/recharge.ts
export const RECHARGE_OPERATORS = [
  "AIRTEL",
  "JIO",
  "VI",
  "BSNL",
  "ALL_IN_ONE",
] as const;

export type RechargeOperator = (typeof RECHARGE_OPERATORS)[number];

export const RECHARGE_AMOUNT_FIELDS = [
  { key: "saleProfit", entryType: "SALE_PROFIT", label: "Sale Profit" },
  { key: "chillar", entryType: "CHILLAR", label: "Chillar" },
  { key: "act", entryType: "ACT", label: "ACT" },
  { key: "mnp", entryType: "MNP", label: "MNP" },
] as const;

export type RechargeAmountFieldKey = (typeof RECHARGE_AMOUNT_FIELDS)[number]["key"];

export const RECHARGE_ENTRY_TYPE_LABELS: Record<string, string> = {
  SALE_PROFIT: "Sale Profit",
  CHILLAR: "Chillar",
  ACT: "ACT",
  MNP: "MNP",
  MULTI: "Multiple",
};

export function getRechargeEntryTypeLabel(entryType: string): string {
  return RECHARGE_ENTRY_TYPE_LABELS[entryType] ?? entryType;
}

export type RechargeBreakdownRow = {
  saleProfit?: string | number | null;
  chillar?: string | number | null;
  act?: string | number | null;
  mnp?: string | number | null;
  entryType?: string;
  amount?: string | number | null;
};

export function getRechargeBreakdownParts(row: RechargeBreakdownRow) {
  const hasColumns = RECHARGE_AMOUNT_FIELDS.some((f) => {
    const v = row[f.key as keyof RechargeBreakdownRow];
    return v != null;
  });

  if (hasColumns) {
    return RECHARGE_AMOUNT_FIELDS.flatMap((f) => {
      const v = row[f.key as keyof RechargeBreakdownRow];
      const n = Number(v ?? 0);
      if (n <= 0) return [];
      return [{ entryType: f.entryType, label: f.label, amount: String(n) }];
    });
  }

  if (row.entryType) {
    const n = Number(row.amount ?? 0);
    return [
      {
        entryType: row.entryType,
        label: getRechargeEntryTypeLabel(row.entryType),
        amount: String(n),
      },
    ];
  }

  return [];
}

export function formatRechargeTypeLabel(row: RechargeBreakdownRow): string {
  const parts = getRechargeBreakdownParts(row);
  return parts.map((p) => p.label).join(" + ") || "—";
}

export function rechargeEntryHasType(row: RechargeBreakdownRow, entryType: string): boolean {
  return getRechargeBreakdownParts(row).some((p) => p.entryType === entryType);
}

File: apps/flutter/lib/domain/constants/recharge.dart
const rechargeOperators = ['AIRTEL', 'JIO', 'VI', 'BSNL', 'ALL_IN_ONE'];

class RechargeAmountField {
  const RechargeAmountField({required this.key, required this.entryType, required this.label});
  final String key;
  final String entryType;
  final String label;
}

const rechargeAmountFields = <RechargeAmountField>[
  RechargeAmountField(key: 'saleProfit', entryType: 'SALE_PROFIT', label: 'Sale Profit'),
  RechargeAmountField(key: 'chillar', entryType: 'CHILLAR', label: 'Chillar'),
  RechargeAmountField(key: 'act', entryType: 'ACT', label: 'ACT'),
  RechargeAmountField(key: 'mnp', entryType: 'MNP', label: 'MNP'),
];

const rechargeEntryTypeLabels = <String, String>{
  'SALE_PROFIT': 'Sale Profit',
  'CHILLAR': 'Chillar',
  'ACT': 'ACT',
  'MNP': 'MNP',
  'MULTI': 'Multiple',
};

String getRechargeEntryTypeLabel(String entryType) =>
    rechargeEntryTypeLabels[entryType] ?? entryType;

class RechargeBreakdownPart {
  RechargeBreakdownPart({required this.entryType, required this.label, required this.amount});
  final String entryType;
  final String label;
  final String amount;
}

List<RechargeBreakdownPart> getRechargeBreakdownParts(Map<String, dynamic> row) {
  final hasColumns = rechargeAmountFields.any((f) => row[f.key] != null);
  if (hasColumns) {
    return rechargeAmountFields.expand((f) {
      final n = double.tryParse('${row[f.key] ?? 0}') ?? 0;
      if (n <= 0) return <RechargeBreakdownPart>[];
      return [RechargeBreakdownPart(entryType: f.entryType, label: f.label, amount: '$n')];
    }).toList();
  }
  if (row['entryType'] != null) {
    final n = double.tryParse('${row['amount'] ?? 0}') ?? 0;
    return [
      RechargeBreakdownPart(
        entryType: '${row['entryType']}',
        label: getRechargeEntryTypeLabel('${row['entryType']}'),
        amount: '$n',
      ),
    ];
  }
  return [];
}

String formatRechargeTypeLabel(Map<String, dynamic> row) {
  final parts = getRechargeBreakdownParts(row);
  return parts.map((p) => p.label).join(' + ').isEmpty ? '—' : parts.map((p) => p.label).join(' + ');
}
